crates/agent/tests/e2e_proposal_one.rs

```diff
@@ -0,0 +1,817 @@
+            }
+        };
+
+        // Use our new direct_query_agent function for better responses
+        let answer = match direct_query_agent(agent_service, &prompt).await {
+            Ok(response) => response,
+            Err(e) => {
+                if retries < MAX_RETRIES {
+                    println!("Error on attempt {}: {:?}. Retrying...", retries + 1, e);
+                    retries += 1;
+                    tokio::time::sleep(Duration::from_millis(2000)).await; // Wait a bit before retrying
+                    continue;
+                } else {
+                    return Err(e.into()); // Convert agent::Error to anyhow::Error
+                }
+            }
+        };
+
+        // If we got a non-empty answer, check if it's valid for binary questions
+        if !answer.trim().is_empty() {
+            // For binary questions, check if the answer starts with yes or no
+            if is_binary
+                && !answer.to_lowercase().starts_with("yes")
+                && !answer.to_lowercase().starts_with("no")
+            {
+                if retries < MAX_RETRIES {
+                    println!(
+                        "Invalid binary response format on attempt {}. Retrying...",
+                        retries + 1
+                    );
+                    retries += 1;
+                    tokio::time::sleep(Duration::from_millis(1000)).await;
+                    continue;
+                }
+            } else {
+                return Ok(answer);
+            }
+        }
+
+        // If we still have an empty or invalid answer and have retries left, try again
+        if retries < MAX_RETRIES {
+            println!("Empty response on attempt {}. Retrying...", retries + 1);
+            last_answer = answer; // Save the last answer in case all attempts fail
+            retries += 1;
+            tokio::time::sleep(Duration::from_millis(2000)).await; // Wait a bit before retrying
+        } else {
+            // We've exhausted our retries, return the best answer we have or a fallback
+            if !last_answer.trim().is_empty() {
+                return Ok(last_answer);
+            }
+
+            // Final fallback
+            if is_binary {
+                // For binary questions, default to "unknown" to fail the test explicitly
+                return Ok("unknown".to_string());
+            } else {
+                return Ok("No relevant information found in the proposal".to_string());
+            }
+        }
+    }
+
+    // This should never be reached due to the returns above
+    Ok(last_answer)
+}
+
+/// Compare expected and actual answers, returning true if they match
+/// For binary questions, only check if the answer starts with yes or no
+/// For reasoning questions, use a more flexible semantic comparison
+fn compare_answers(expected: &str, actual: &str, is_binary: bool) -> bool {
```

Comment:
Consider splitting this function into atomic functions

Reply:
https://github.com/NethermindEth/wei/pull/39/commits/28a05870be1ec74f72a5b31341f71b51c906b4e3